### PR TITLE
Support Loading package.json from parent dir

### PR DESCRIPTION
**Pull request by RobbieMcKinstry**
# Description

This PR changes how the Node SDK resolves `package.json` to recursively look up the directory tree for the file when it's
not present in the cwd. The previous behavior was to only look in the current directory, and fallback to an empty object
when the file could not be found.

Fixes https://github.com/pulumi/pulumi/issues/10459, https://github.com/pulumi/pulumi/pull/11045,
and https://github.com/pulumi/pulumi/issues/2619

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Comment by RobbieMcKinstry** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12758** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12758" target="_blank">Graphite</a> 
    * **PR #12759** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12759" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/12759?utm_source=stack-comment).

**Comment by RobbieMcKinstry** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12758** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12758" target="_blank">Graphite</a> 
    * **PR #12759** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/12759" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/12759?utm_source=stack-comment).

**Comment by RobbieMcKinstry** [comment]
bors try

File: changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml
Comment (abhinav) [inline changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml:4]: wording suggestions:

```suggestion
  description: Support loading package.json from parent directory. If `package.json` is not found in the current directory, Pulumi recursively searches up the directory tree until it is found. If `package.json` provides a `main` field, that field is relative to the current directory.
```

Comment (abhinav) [reply changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml:4]: More importantly:

> If `package.json` provides a `main` field, that field is relative to the current directory

Isn't that supposed to be relative to the directory containing package.json, not the current directory?
Is the behavior incorrect or the description?

File: sdk/nodejs/cmd/run/run.ts
Comment (abhinav) [inline sdk/nodejs/cmd/run/run.ts:74]: minor: Mind running the formatter over this? `if(` here versus `if (` below.

File: sdk/nodejs/cmd/run/run.ts
Comment (abhinav) [inline sdk/nodejs/cmd/run/run.ts:67]: I appreciate the surrounding comments. Would you mind adding a comment above this function about what programPath is?

File: sdk/nodejs/cmd/run/run.ts
Comment (abhinav) [inline sdk/nodejs/cmd/run/run.ts:83]: nit: s/pulumi/Pulumi
```suggestion
        log.warn("Could not find a package.json file for the program. Using the Pulumi program directory as the project root.");
```

File: sdk/nodejs/cmd/run/run.ts
Comment (abhinav) [inline sdk/nodejs/cmd/run/run.ts:98]: I'm not sure I understand this change?
The semantics of return from a `finally` are non-obvious to me, but if I'm reading this correctly, this will behave as follows:

- if there is no error, return the result of `require(packageJson)`
- if there is an error, it will return `{}` because `packageObject` is empty

If that's the behavior this is aiming for, the old version has the same behavior, and is more obvious, IMO.

File: tests/integration/integration_nodejs_test.go
Comment (abhinav) [inline tests/integration/integration_nodejs_test.go:1147]: I think you can probably `t.Parallel` these two tests.

File: tests/integration/read/read_replace/step1/package.json
Comment (abhinav) [inline tests/integration/read/read_replace/step1/package.json:4]: Could you clarify why we need these changes?

Comment (RobbieMcKinstry) [reply tests/integration/read/read_replace/step1/package.json:4]: Sorry, I wrote a comment to clarify when we needed these changes, but I never hit submit! lol You read my mind XD

https://github.com/pulumi/pulumi/pull/12759#discussion_r1179654521

File: changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml
Comment (RobbieMcKinstry) [reply changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml]: > Isn't that supposed to be relative to the directory containing package.json, not the current directory?

So I thought about this a bit when writing the test cases, and I believe the answer is "no", but I definitely want to hear your thoughts.

The most common use-case for putting `package.json` above pwd is a monorepo, where that `package.json` is serving multiple Pulumi projects. Suppose that `package.json` has a `main` field. That field will always point to the same entrypoint, no matter which Pulumi project is running. 

Instead, if `main` is relative to `pwd`, then you can universally rename your entrypoint. For example, suppose you wanted your entrypoint to be `main.ts` in every Pulumi project. But providing `"main": "main.ts"` in your `package.json`, you will override the default `index.js` for all Pulumi projects in the monorepo. I think that's the only time this would make sense.

----------

This is what you'd have if `main` is relative to `pwd`.
```
.
└── my-monorepo/
    ├── package.json
    ├── node_modoules/
    │   ├── @pulumi/pulumi
    │   └── ...
    ├── project-A/
    │   ├── Pulumi.yaml
    │   └── main.ts
    └── project-B/
        ├── Pulumi.yaml
        └── main.ts
```

This is what you'd have if main is relative to `package.json`
```
.
└── my-monorepo/
    ├── package.json
    ├── main.ts
    ├── node_modoules/
    │   ├── @pulumi/pulumi
    │   └── ...
    ├── project-A/
    │   └── Pulumi.yaml
    └── project-B/
        └── Pulumi.yaml
```

This precludes monorepos from using this feature meaningfully.

File: sdk/nodejs/cmd/run/run.ts
Comment (RobbieMcKinstry) [reply sdk/nodejs/cmd/run/run.ts:74]: Am i losing my mind? I'm not seeing a formatted or formatter config in this directory.
We don't have a `make format` rule for JavaScript, nor is `prettier` installed as a dev dependency. The `make format` rule just runs gofmt. 😕 ❓ 

I can open an issue to add a formatter to the repo.

For now I just added the space, which looks consistent with other locations in the file. 🤷🏻

File: sdk/nodejs/cmd/run/run.ts
Comment (RobbieMcKinstry) [reply sdk/nodejs/cmd/run/run.ts:98]: I'll revert this change. I felt it was a style improvement to have a single return in this function. Your description of the behavior is correct; it's a refactor only. 👍🏻

File: tests/integration/integration_nodejs_test.go
Comment (RobbieMcKinstry) [reply tests/integration/integration_nodejs_test.go]: Done!

File: changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml
Comment (abhinav) [reply changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml]: Thanks for the `@`, @RobbieMcKinstry otherwise I would have missed this.
I see what you're trying to do with the UX here, but I don't think we have that flexibility here.
Per the [NPM docs](https://docs.npmjs.com/cli/v6/configuring-npm/package-json#main)

> This should be a module ID relative to the root of your package folder.

We cannot change that behavior—users expect main to be relative to the package directory—the one containing the package.json

But you're right, this exposes a big gap in this feature.
Moving the package.json up one level still won't allow users to have multiple Pulumi programs coexist in the same NPM package, or even a normal program and a Pulumi program in the same package.
We need to find a different dial for that, though. As mentioned above, I don't think we can change the meaning of what `"main"` means in the package.json.
One option I can imagine is a field for the JS backend in the Pulumi.yaml that specifies the script to run, defaulting to the `main` specified in the package.json. That value can be relative to the Pulumi.yaml directory.

File: sdk/nodejs/cmd/run/run.ts
Comment (RobbieMcKinstry) [reply sdk/nodejs/cmd/run/run.ts]: https://github.com/pulumi/pulumi/issues/12781

File: changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml
Comment (RobbieMcKinstry) [reply changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml]: Ok, I'll get this to work using main relative to package.json instead of pwd to adhere to the spec. :)

This feature is still useful even if main has to be relative to package.json. This will be an improvement for anyone in a monorepo using index.js or index.ts as an entry point, which is the default across all Node projects, in addition to the subset which is Pulumi Node projects.

File: changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml
Comment (Frassle) [inline changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml:4]: I feel "Relative to the current directory" isn't precise enough. 
Relative to the directory the package.json was found?
Relative to the current working directory pulumi was started in?
Relative to the project root (i.e. where Pulumi.yaml was found)?
Relative to the project main (i.e. where "main" in the Pulumi.yaml pointed)?

File: sdk/nodejs/cmd/run/run.ts
Comment (Frassle) [inline sdk/nodejs/cmd/run/run.ts:72]: Not "the current working directory", this is a mistake we keep making 😆 
It should be the project "main" directory, which defaults to project "root" directory.

File: sdk/nodejs/cmd/run/run.ts
Comment (Frassle) [inline sdk/nodejs/cmd/run/run.ts:389]: s/working/main

File: sdk/nodejs/cmd/run/run.ts
Comment (RobbieMcKinstry) [reply sdk/nodejs/cmd/run/run.ts:72]: Good point!

File: tests/integration/integration_nodejs_test.go
Comment (RobbieMcKinstry) [inline tests/integration/integration_nodejs_test.go:1136]: I believe the reason integration tests aren't marked as parallel elsewhere in the file is because they are automatically marked as parallel in `integration.ProgramTest`

File: changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml
Comment (RobbieMcKinstry) [reply changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml]: @abhinav I've updated the tests to reflect the `npm` standard behavior. This should be more than sufficient. I've learned that if you're using a monorepo, you're probably _**not**_ using `main`, because that would defeat the purpose.

**Comment by RobbieMcKinstry** [comment]
@abhinav I've updated the tests to reflect the `npm` standard behavior. This should be more than sufficient. I've learned that if you're using a monorepo, you're probably _**not**_ using `main`, because that would defeat the purpose. 

Waiting on confirmation that the tests pass; everything looks good locally but I have little faith in CI XD After that, this should be good to merge.

File: changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml
Comment (RobbieMcKinstry) [reply changelog/pending/20230426--sdk-nodejs--support-loading-package-json-from-parent-directory.yaml]: I addresses this by invoking the terminology you've used elsewhere.

**Review by Frassle — APPROVED** [review]
Looks reasonable to me

File: tests/integration/aliases/nodejs/adopt_into_component/step1/package.json
Comment (justinvp) [reply tests/integration/aliases/nodejs/adopt_into_component/step1/package.json:5]: What will be the impact for users if they have a `main` leftover in their `package.json` that isn't currently used? Do we need to call out anything else explicitly in the changelog so users are aware?

File: sdk/nodejs/cmd/run/run.ts
Comment (abhinav) [inline sdk/nodejs/cmd/run/run.ts:399]: I'm guessing these log statements are just printed as diagnostics if there are issues?

**Review by abhinav — APPROVED** [review]

**Review by justinvp — APPROVED** [review]
One question, but otherwise LGTM. Please squash commits before merging.

File: sdk/nodejs/cmd/run/run.ts
Comment (RobbieMcKinstry) [reply sdk/nodejs/cmd/run/run.ts:399]: Whoops, those snuck in there today! I'll of course rip these out before whacking merging!

File: tests/integration/aliases/nodejs/adopt_into_component/step1/package.json
Comment (RobbieMcKinstry) [reply tests/integration/aliases/nodejs/adopt_into_component/step1/package.json:5]: @justinvp I've already called this out in the changelog, but since you and others have asked, it's worth mentioning this:

1. Largely, users with a `package.json` outside the Pulumi root directory could not run their programs before this change. So nobody who currently has a working program would be affected by this change.
2. This change increases the set of programs that Pulumi accepts to include those with a `package.json` outside of the Pulumi root directory. For these programs, we use the behavior [as defined by NPM](https://docs.npmjs.com/cli/v6/configuring-npm/package-json#main). (Yarn [does not offer guidance](https://classic.yarnpkg.com/lang/en/docs/package-json/#toc-main).) In short, **the `main` file is relative to the directory containing `package.json`.** Previously, this was the same as saying the Pulumi `root` directory.

File: sdk/nodejs/cmd/run/run.ts
Comment (RobbieMcKinstry) [reply sdk/nodejs/cmd/run/run.ts:399]: They were added for debugging a round of tests earlier today! XD

File: sdk/nodejs/cmd/run/run.ts
Comment (RobbieMcKinstry) [reply sdk/nodejs/cmd/run/run.ts:85]: I think that's a good change. While nobody loves exception-based-control-flow, the sync-in-async issue feels more insidious and easier to miss. Thanks! :D

**Comment by RobbieMcKinstry** [comment]
Thanks, everyone! Lots of good feedback here :)

**Pull request merged**
